// helper/index.js
/**
 * Created with JetBrains WebStorm.
 * User: 松松
 * Date: 13-9-4
 * Time: 上午11:05
 * To change this template use File | Settings | File Templates.
 */

var app = require('app')
var db = require('db')
var ObjectID = db.mongodb.ObjectID
var path = require('path')

exports.csrf = function (req, res, next) {
    res.locals.token = req.csrfToken();
    next();
}

//判断当前是否为登陆状态
exports.isLogin = function (req) {
    return req.session._id !== undefined;
}

//检测某个用户是否拥有当前角色的权限

exports.isGroup = function (req, group, fn) {
    try {
        var id = ObjectID(req.session._id)
    } catch (e) {
        fn()
    }

    var user = new db.mongodb.Collection(db.userClient, 'user')
    user.findOne({_id: id, status: {$gt: 0}, group: {$in: group}}, {_id: 1}, function (err, docs) {
        fn(!err && docs && docs._id)
    })
}

//获取登陆成员的组角色
exports.getGroup = function (req, fn) {
    try {
        var id = ObjectID(req.session._id)
    } catch (e) {
        fn()
    }

    var user = new db.mongodb.Collection(db.userClient, 'user')
    user.findOne({_id: id, status: {$gt: 0}}, {_id: 1, group: 1}, function (err, docs) {
        if (!err && docs && docs.group) {
            fn(docs.group)
        } else {
            fn()
        }
    })
}

//add notice
exports.addUserSessionNotice = function (req, key, value) {
    if (!req.session) return
    if (!req.session.userSessionNotice) req.session.userSessionNotice = Object.create(null)
    req.session.userSessionNotice[key] = value
}

//remove notice
exports.removeUserSessionNotice = function (req, key) {
    if (req.session && req.session.userSessionNotice) {
        delete req.session.userSessionNotice[key]
    }
}
