// user-indicators/index.js
/**
 * Created by 松松 on 13-11-11.
 */

/*
 更新用户的作品数和文章数量
 * */

var db = require('db')
exports.update = function (id, collection) {

    try {
        var _id = db.mongodb.ObjectID(id)
    } catch (e) {
        console.log(e)
        return
    }


    var map = {
        'design-works': {
            filter: {
                owner_id: id,
                //只统计个人类型的作品
                type: 'own',
                status: { $gte: 1 }
            }
        },
        article: {
            filter: {
                owner_id: id,
                status: { $gte: 1 }
            }
        }
    }

    if (!map[collection]) {
        console.log('无法更新' + collection + '的指标', __filename)
        return
    }

    var col = new db.mongodb.Collection(db.Client, collection)

    col.count(map[collection].filter, function (err, count) {
        if (!err) {
            var field = {$set: {}}
            //更新对应的指标
            field.$set['index.' + collection] = count
            var user = new db.mongodb.Collection(db.userClient, 'user')
            user.update({_id: _id}, field, {w: 1}, function (err, result) {
            })
        }
    })
}
